### PR TITLE
options: add build-jobs

### PR DESCRIPTION
**Pull request by kamilrakoczy**
Pre-PR to: https://github.com/verilator/verilator/pull/3608

This PR adds ``--build-jobs`` options that controls parallelism for ``--build``. ``-j`` option still sets ``--build-jobs`` value, but in the future, this option will also set ``--verilation-jobs``.

Signed-off-by: Kamil Rakoczy <krakoczy@antmicro.com>

### CODE REVIEW

**Comment by gezalore** [comment]
We want --build-jobs to default to the value of -j if not explicitly given, that is: `--build-jobs 2 -j 10` should build on 2 threads. This will not work with the current implementation. For some reason `-j` is parsed at the bottom separately, probably should just put that into the standard section unless there is some strange reason it is where it is (there shouldn't be).

File: docs/guide/exe_verilator.rst [hunk 1]
```diff
@@ -129,6 +129,12 @@ Summary:
    is also used). Verilator manages the build itself, and for this --build
    requires GNU Make to be available on the platform.
 
+.. option:: --build-jobs [<value>]
+
+   Specify the level of parallelism for :vlopt:`--build`. The <value> must
+   be a positive integer specifying the maximum number of parallel build
+   jobs.
```
Comment (wsnyder) [inline docs/guide/exe_verilator.rst:136]:
```suggestion
   jobs.
   
   See also :vlopt:`-j`.
```

**Review by wsnyder — COMMENTED** [review]
>We want --build-jobs to default to the value of -j if not explicitly given, that is: --build-jobs 2 -j 10 should build on 2 threads
That's different from my suggestion but probably better, specifically you'd default buildJobs to 0, then in ::notify have e.g.
 if (!m_buildJobs) m_buildJobs = m_jobs.
  if (m_buildJobs == 1) m_buildJobs = 0;
Note we want "-j 10 --verilation-jobs 1" to mean no parallelism.

File: docs/guide/exe_verilator.rst (hunk 1 above)
Comment (wsnyder) [reply docs/guide/exe_verilator.rst:136]: Also update docs for -j to say this is default for --build-jobs if --build-jobs isn't provided.

**Comment by kamilrakoczy** [comment]
I've changed default ``m_buildJobs`` to 0 and now I only set ``m_buildJobs`` if it is 0 as @wsnyder suggested, but this changed default behavior when there is no ``--build-jobs`` and ``-j`` flag. Now it defaults to unlimited (0) where before it was 1.

I think I would need to introduce extra variable to keep track if some flag was seen if we don't want to change this behavior.

**Comment by gezalore** [comment]
> I think I would need to introduce extra variable to keep track if some flag was seen if we don't want to change this behavior.

Can't you use a different sentinel value? -1, or numerical_limits::max() ?

If explicitly set, done
Else default to -j if -j given
Else default to no parallelism.

**Comment by gezalore** [comment]
FWIW I would also be happy not to have the "unlimited parallelism" option (i.e.: no naked `-j` without numeric argument), as it's usually a bad idea. User can always just do `-j $(nproc)`

**Comment by wsnyder** [comment]
For better and worse "make -j" (without number) has a standard meaning I'd prefer to support.

File: src/V3Options.cpp [hunk 2]
```diff
@@ -1510,11 +1515,13 @@ void V3Options::parseOptsList(FileLine* fl, const string& optdir, int argc, char
...
-                if (m_buildJobs <= 0) {
-                    fl->v3error("-j accepts positive integer, but '" << argv[i] << "' is passed");
+                if (m_buildJobs < 1) {
+                    m_buildJobs = atoi(argv[i]);
+                    if (m_buildJobs <= 0) {
+                        fl->v3error("-j accepts positive integer, but '" << argv[i] << "' is passed");
+                    }
                 }
                 ++i;
             }
```
[the remark this answers is not in the document]
Comment (wsnyder) [reply src/V3Options.cpp:1527]: And after the error set val = 1;

**Comment by gezalore** [comment]
So are we saying naked `-j` will run both verilation *and the build* with `std::thread::hardware_concurrency()`?

File: src/V3Options.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (gezalore) [reply src/V3Options.cpp]: Note: also tweaked the English in the message, please apply with `--build-jobs` as well.

**Comment by gezalore** [comment]
See:
https://github.com/verilator/verilator/issues/2981
https://github.com/verilator/verilator/issues/3139

**Comment by wsnyder** [comment]
How about this then, I can yield on "-j" standalone.

none -> set verilation-threads to unthreaded, pass no -j to make.
"-j 0" -> set verilation-threads to HW_concurrency, pass -j HW_concurrency to make.
"-j 1" -> set verilation-threads to unthreaded, pass -j 1 to make.
"-j 2" -> set verilation-threads to 2, pass -j 2 to make.

These apply only if no --verilation-threads or --build-threads passed (so need -1 or another magic number internally).  Those options would have similar three meanings above.

**Comment by gezalore** [comment]
> How about this then, I can yield on "-j" standalone.
> 
> none -> set verilation-threads to unthreaded, pass no -j to make. "-j 0" -> set verilation-threads to HW_concurrency, pass -j HW_concurrency to make. "-j 1" -> set verilation-threads to unthreaded, pass -j 1 to make. "-j 2" -> set verilation-threads to 2, pass -j 2 to make.
> 
> These apply only if no --verilation-threads or --build-threads passed (so need -1 or another magic number internally). Those options would have similar three meanings above.

Sounds good to me.

Is there a difference between `make` and `make -j 1` ? if no, just do the same with both "verialtor <no-j-option>" and "verilator -j 1" for simplicity of implementation.

**Comment by wsnyder** [comment]
>Is there a difference between make and make -j 1 ?

Yes, a submake might get passed a "-j other" which this may override.  In theory we are creating the makefile so that might not apply, but I think it's better not to add flags the user didn't add themselves somehow.

**Comment by kamilrakoczy** [comment]
> How about this then, I can yield on "-j" standalone.
> 
> none -> set verilation-threads to unthreaded, pass no -j to make. "-j 0" -> set verilation-threads to HW_concurrency, pass -j HW_concurrency to make. "-j 1" -> set verilation-threads to unthreaded, pass -j 1 to make. "-j 2" -> set verilation-threads to 2, pass -j 2 to make.
> 
> These apply only if no --verilation-threads or --build-threads passed (so need -1 or another magic number internally). Those options would have similar three meanings above.

I've updated PR to match this description. I've also deleted ``test_regress/t/t_flag_j_bad`` test as now we are accepting ``0`` as possible value and ``-1`` is interpreted as another option, not negative value, so I couldn't test this.

**Comment by gezalore** [comment]
> How about this then, I can yield on "-j" standalone.

I think this meant remove naked `-j` @wsnyder ?

> I've updated PR to match this description. I've also deleted test_regress/t/t_flag_j_bad test as now we are accepting 0 as possible value and -1 is interpreted as another option, not negative value, so I couldn't test this.

We should still not accept `-j 0123SNAFU` which would fail that test.

**Comment by kamilrakoczy** [comment]
> We should still not accept `-j 0123SNAFU` which would fail that test.

``atoi`` function skips non-numeric values after valid number representation, see: https://en.cppreference.com/w/c/string/byte/atoi (https://godbolt.org/z/Monz9rhdx), so we would convert this to ``123``.

Anyway, if this description meant that we want to remove naked ``-j``, I will move handle of ``-j`` to standard section and than we can test for ``-1``.

**Comment by gezalore** [comment]
> `atoi` function skips non-numeric values after valid number representation, see: https://en.cppreference.com/w/c/string/byte/atoi (https://godbolt.org/z/Monz9rhdx), so we would convert this to `123`.

That's why we should use `strtoul` and friends instead, and check the whole argument was consumed:
https://en.cppreference.com/w/c/string/byte/strtoul

https://github.com/gezalore/verilator/blob/94e8770085238c179c5462eecb74bbab8d98933f/src/V3Os.cpp#L117-L124

Accepting cruft is not a good thing.

**Comment by kamilrakoczy** [comment]
I agree, but maybe lets leave this to another PR as ``atoi`` is used in many places in ``V3Options``.

File: docs/guide/exe_verilator.rst [hunk 3]
```diff
@@ -129,6 +129,14 @@ Summary:
...
```
Comment (wsnyder) [inline docs/guide/exe_verilator.rst:136]:
```suggestion
.. option:: --build-jobs <value>

   Specify the level of parallelism for :vlopt:`--build`. If zero, uses the
   number of threads in the current hardware. Otherwise, the <value> must
   be a positive integer specifying the maximum number of parallel build
   jobs.
```

And fix -j similar.

File: src/Verilator.cpp [hunk 4]
```diff
@@ -663,11 +663,7 @@ static string buildMakeCmd(const string& makefile, const string& target) {
     cmd << v3Global.opt.getenvMAKE();
     cmd << " -C " << v3Global.opt.makeDir();
     cmd << " -f " << makefile;
-    if (jobs == 0) {
-        cmd << " -j";
-    } else if (jobs > 1) {
-        cmd << " -j " << jobs;
-    }
+    if (jobs > 0) { cmd << " -j " << jobs; }
```
Comment (wsnyder) [inline src/Verilator.cpp:666]:
```suggestion
    if (jobs > 0) cmd << " -j " << jobs;
```

**Comment by wsnyder** [comment]
@kamilrakoczy Not sure why didn't show up on pull's run, but there's a regression failure for this on 18.04 - please make a separate pull to fix it, thanks.

https://github.com/verilator/verilator/actions/runs/3060442052/jobs/4939552768

**Comment by gezalore** [comment]
> @kamilrakoczy Not sure why didn't show up on pull's run, but there's a regression failure for this on 18.04 - please make a separate pull to fix it, thanks.

PRs only run on 20.04

**Comment by kamilrakoczy** [comment]
So the problem is that ``make`` added information about number of jobs to ``MAKEFLAGS`` in 4.2 (https://lists.gnu.org/archive/html/info-gnu/2016-05/msg00013.html), but ubuntu 16.04 still uses 4.1. Looking for a way to workaround this limitation.

**Comment by kamilrakoczy** [comment]
I've created PR that fixes failing test (I've checked it in local docker container) by adding build jobs information to stats file and checking for correct value: https://github.com/verilator/verilator/pull/3626
